digdag-core/src/main/java/io/digdag/core/database/DatabaseMigrator.java

```diff
@@ -626,11 +626,117 @@ public void migrate(Handle handle)
+                    " from queued_tasks");
+
+            // queues
+            handle.update("alter table queues" +
+                    " add column shared_site_id int");
+
+            // resource_types
+            handle.update("drop table resource_types");
+
+            // queued_tasks
+            handle.update("alter table queued_tasks" +
+                    " alter column task_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " alter column queue_id drop not null");
+            handle.update("alter table queued_tasks" +
+                    " drop column priority");
+            handle.update("alter table queued_tasks" +
+                    " drop column resource_type_id");
+            handle.update("create unique index queued_tasks_on_site_id_task_id on queued_tasks (site_id, task_id)");
+
+            if (isPostgres()) {
+                handle.update(
+                    "CREATE FUNCTION lock_shared_tasks(target_site_id int, target_site_max_concurrency bigint, limit_count int, lock_expire_seconds int, agent_id text) returns setof bigint as $$\n" +
+                    "BEGIN\n" +
+                    "  IF pg_try_advisory_xact_lock(23300, target_site_id) THEN\n" +
+                    "    RETURN QUERY\n" +
+                    "      with updated as (\n" +
+                    "        update queued_task_locks\n" +
+                    "        set lock_expire_time = cast(extract(epoch from statement_timestamp()) as bigint) + lock_expire_seconds,\n" +
+                    "            lock_agent_id = agent_id\n" +
+                    "        where id = any(\n" +
+                    "          select queued_task_locks.id\n" +
+                    "          from queued_task_locks\n" +
+                    "          where lock_expire_time is null\n" +
+                    "          and site_id = target_site_id\n" +
+                    "          and not exists (\n" +
+                    "            select * from (\n" +
+                    "              select queue_id, count(*) as count\n" +
+                    "              from queued_task_locks\n" +
+                    "              where lock_expire_time is not null\n" +
+                    "                and site_id = target_site_id\n" +
+                    "              group by queue_id\n" +
+                    "            ) runnings\n" +
+                    "            join queues on queues.id = runnings.queue_id\n" +
+                    "            where runnings.count >= queues.max_concurrency\n" +
+                    "              and runnings.queue_id = queued_task_locks.queue_id\n" +
+                    "          )\n" +
+                    "          and not exists (\n" +
+                    "            select count(*)\n" +
+                    "            from queued_task_locks\n" +
+                    "            where lock_expire_time is not null\n" +
```

Comment:
How?
This subquery doesn't depend on outer query. Thus PostgreSQL compiles this subquery to a InitPlan instead of a anti semi join. If this InitPlan returns 1 or more rows (meaning that target site already exceeds max_concurrency), all following plans won't be executed at all.